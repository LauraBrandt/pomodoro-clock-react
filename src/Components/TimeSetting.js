import React from 'react';

const TimeSetting = ({type, length}) => {
  return(
    <div>
      <label htmlFor={`${type && type.toLowerCase()}-length`} id={`${type && type.toLowerCase()}-label`}>{type} Length</label>
      <div>
        <button id={`${type && type.toLowerCase()}-decrement`}><i className="fas fa-minus" title="minus"></i></button>
        <input 
          id={`${type && type.toLowerCase()}-length`} 
          type="number" 
          value={length}
        />
        <button id={`${type && type.toLowerCase()}-increment`}><i className="fas fa-plus" title="plus"></i></button>
      </div>
    </div>
  );
}

export default TimeSetting;