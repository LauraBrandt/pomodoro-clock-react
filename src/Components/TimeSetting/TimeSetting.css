.time-setting label {
  display: block;
  text-align: center;
  padding-bottom: 1rem;
  font-size: 1.1rem;
}

.length-controls {
  display: flex;
  justify-content: center;
}

.length-controls button {
  border: none;
  background: #222;
  color: white;
  font-size: 1.2rem;
  cursor: pointer;
  padding: 0 0.3rem;
  margin: 0 0.4rem;
}

.length-controls input {
  border: none;
  width: 3rem;
  text-align: center;
  background: #222;
  color: white;
  font-size: 1.5rem;
  padding: 0;
  margin: 0;
}

/* Remove number spinner */
.length-controls input[type='number']::-webkit-outer-spin-button,
.length-controls input[type='number']::-webkit-inner-spin-button {
    -webkit-appearance: none;
    margin: 0;
}
.length-controls input[type='number'] {
  -moz-appearance:textfield;
}

.length-controls input:focus,
.length-controls button:focus{
  outline: none;
  background: #333;
  border-radius: 5px;
}

@media (max-width: 450px) {
  .time-setting {
    margin: 0.8em 0;
  }
}