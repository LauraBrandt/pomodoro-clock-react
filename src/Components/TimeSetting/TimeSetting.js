import React from 'react';
import PropTypes from 'prop-types';
import './TimeSetting.css';

const TimeSetting = ({type, length, handleChangeByOne, handleChange}) => {
  return(
    <div className="time-setting">
      <label htmlFor={`${type && type.toLowerCase()}-length`} id={`${type && type.toLowerCase()}-label`}>{type} Length</label>
      <div className="length-controls">
        <button id={`${type && type.toLowerCase()}-decrement`} onClick={handleChangeByOne.bind(this, type, -1)}>
          <i className="fas fa-minus" title="minus"></i>
        </button>
        <input 
          id={`${type && type.toLowerCase()}-length`} 
          type="number" 
          value={length}
          onChange={handleChange}
          onBlur={handleChange}
          min="0"
          max="60"
        />
        <button id={`${type && type.toLowerCase()}-increment`}  onClick={handleChangeByOne.bind(this, type, 1)}>
          <i className="fas fa-plus" title="plus"></i>
        </button>
      </div>
    </div>
  );
}

TimeSetting.propTypes = {
  type: PropTypes.string.isRequired,
  length: PropTypes.number.isRequired,
  handleChangeByOne: PropTypes.func.isRequired,
  handleChange: PropTypes.func.isRequired
}

export default TimeSetting;