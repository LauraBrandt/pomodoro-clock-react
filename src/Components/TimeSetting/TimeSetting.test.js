import React from 'react';
import {shallow, mount} from 'enzyme';
import renderer from 'react-test-renderer';
import TimeSetting from './TimeSetting';

it('renders without crashing', () => {
  shallow(<TimeSetting type="" length={25} handleChangeByOne={jest.fn()} handleChange={jest.fn()} />);
});

describe('structure', () => {
  let wrapper = mount(<TimeSetting type="Session" length={25} handleChangeByOne={jest.fn()} handleChange={jest.fn()} />);
  
  it('has all elements with ids necessary', () => {
    expect(wrapper.find('#session-label')).toHaveLength(1);
    expect(wrapper.find('#session-decrement')).toHaveLength(1);
    expect(wrapper.find('#session-length')).toHaveLength(1);
    expect(wrapper.find('#session-increment')).toHaveLength(1);
    expect(wrapper.find('.fa-minus')).toHaveLength(1);
    expect(wrapper.find('.fa-plus')).toHaveLength(1);
  });
  
  it('sets the label correctly', () => {
    expect(wrapper.find('#session-label').text()).toEqual('Session Length');
  });
  
  it('sets the default session time correctly', () => {
    const value = wrapper.find('#session-length').instance().value;
    expect(value).toEqual("25");
  });
  
  it('sets the default break time correctly', () => {
    wrapper.setProps({type: "break", length: 5});
    const value = wrapper.find('#break-length').instance().value;
    expect(value).toEqual("5");
  });
});

describe('event listeners', () => {
  let wrapper, mockHandleChange, mockHandleChangeByOne;
  beforeEach(() => {
    mockHandleChange = jest.fn();
    mockHandleChangeByOne = jest.fn();
    wrapper = shallow(<TimeSetting type="Session" length={25} handleChangeByOne={mockHandleChangeByOne} handleChange={mockHandleChange} />);
  });

  it('the function handleChange is called when the input is changed', () => {
    wrapper.find('#session-length').simulate('change');
    expect(mockHandleChange).toBeCalled();
  });
  
  it('the function handleChange is called when the input is blurred', () => {
    wrapper.find('#session-length').simulate('blur');
    expect(mockHandleChange).toBeCalled();
  });

  it('the function handleChangeByOne is called (with correct args) when the increment button is clicked', () => {
    wrapper.find('#session-increment').simulate('click');
    expect(mockHandleChangeByOne.mock.calls[0][0]).toBe('Session');
    expect(mockHandleChangeByOne.mock.calls[0][1]).toBe(1);
  });
  
  it('the function handleChangeByOne is called (with correct args) when the decrement button is clicked', () => {
    wrapper.find('#session-decrement').simulate('click');
    expect(mockHandleChangeByOne.mock.calls[0][0]).toBe('Session');
    expect(mockHandleChangeByOne.mock.calls[0][1]).toBe(-1);
  });
});

describe('snapshot', () => {
  it('renders', () => {
    const component = renderer.create(<TimeSetting type="Session" length={25} handleChangeByOne={jest.fn()} handleChange={jest.fn()} />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
