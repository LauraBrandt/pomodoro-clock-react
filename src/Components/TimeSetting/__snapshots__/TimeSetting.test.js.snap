// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot renders 1`] = `
<div
  className="time-setting"
>
  <label
    htmlFor="session-length"
    id="session-label"
  >
    Session
     Length
  </label>
  <div
    className="length-controls"
  >
    <button
      id="session-decrement"
      onClick={[Function]}
    >
      <i
        className="fas fa-minus"
        title="minus"
      />
    </button>
    <input
      disabled={false}
      id="session-length"
      max="60"
      min="0"
      onBlur={[Function]}
      onChange={[Function]}
      type="number"
      value={25}
    />
    <button
      id="session-increment"
      onClick={[Function]}
    >
      <i
        className="fas fa-plus"
        title="plus"
      />
    </button>
  </div>
</div>
`;
